### PR TITLE
Skip reformatting and calling onChanged for composing region only changes.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

If a text input formatter keeps rejecting changes made by the IME, and the IME keeps rejecting changes made by the input formatter, then an infinite loop will ensue.  https://github.com/flutter/flutter/pull/66864 made it so that input formatters are reapplied whenever the composing region changes. It turns out Gboard always tries to restore the composing region by calling `setComposingRegion` when the composing region is modified programmatically. If an input formatter removes the composing region, Gboard will try adding it back.

Stop formatting or calling `onChanged` on composing region only changes seems to fix the problem for Gboard (unless the developer tries to do this from within a `TextEditingController`'s listener), for now.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from our [issue database]. Indicate, which of these issues are resolved or fixed by this PR. There should be at least one issue listed here.*

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2177,14 +2186,18 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
 
   void _formatAndSetValue(TextEditingValue value) {
-    // Check if the new value is the same as the current local value, or is the same
-    // as the pre-formatting value of the previous pass (repeat call).
-    final bool textChanged = _value.text != value.text || _value.composing != value.composing;
+    // Only apply input formatters if the text has changed (including uncommited
+    // text in the composing region), or when the user committed the composing
+    // text.
+    // Gboard is very persistent in restoring the composing region. Applying
+    // input formatters on composing region only changes (except clearing the
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2193]: Nit: "composing-region-only changes" if you think that's better.  I misread it and thought that "changes" was a verb at first.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -2177,14 +2186,18 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    // current composing region) is very infinite-loop-prone: the formatters
+    // will keep trying to modify the composing region while Gboard will keep
+    // trying to restore the original composing region.
+    final bool textChanged = _value.text != value.text
+                          || (!_value.composing.isCollapsed && value.composing.isCollapsed);
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2198]: What will happen if a developer does programmatically change the composing region now?  Will the IME still receive the update somehow or will it think that it's using the old composing region?  Or are we saying that developers shouldn't do that (only in TextInputFormatters)?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just curious, did this bug actually happen to anyone or did you just realize it could be a problem yourself?

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2198]: Yeah it's still possible to create an infinite loop in a `TextEditingController` listener callback, but that's like the developer directly fighting the IME, I'm not sure if there's anything we can do to stop that from happening.

**Comment by LongCatIsLooong** [comment]
I was playing with text formatters on master and this bug is very easy to reproduce.

**Pull request merged**